File: server/risk-level.ts
/**
 * Convert per-category model scores into a simple user-facing risk level.
 *
 * The model scores are independent, uncalibrated category scores. This module
 * therefore compares each score with that category's tuned decision threshold
 * from `risk_config.json`; it never presents a raw score as a legal-risk
 * probability.
 *
 * Ported from `bert_multilabel/ml_service/risk_level.py`.
 */

export const HIGH_CONFIDENCE_MARGIN = 0.15
const REVIEW_MARGIN = 0.08

export type RiskLevel = 'low' | 'medium' | 'high'

export type RiskLevelResult = {
  riskLevel: RiskLevel
  riskLevelMessage: string
  detectedCategories: string[]
  reviewCategories: string[]
}

/**
 * Return a risk-level payload for one clause.
 *
 * `high` means either two or more categories were detected, or one score
 * clearly exceeded its own threshold. `medium` means at least one category
 * was detected, or a score is close enough to a threshold to review. `low`
 * means no category was detected and none is close to its threshold.
 */
export function classifyRiskLevel(
  scores: Record<string, number>,
  thresholds: Record<string, number>,
  {
    reviewMargin = REVIEW_MARGIN,
    highConfidenceMargin = HIGH_CONFIDENCE_MARGIN,
  }: { reviewMargin?: number; highConfidenceMargin?: number } = {},
): RiskLevelResult {
  const unknown = Object.keys(scores).filter((category) => !(category in thresholds))
  if (unknown.length) {
    throw new Error(`Missing thresholds for categories: ${unknown.sort().join(', ')}`)
  }
  if (reviewMargin < 0 || highConfidenceMargin < 0) {
    throw new Error('Risk-level margins must be non-negative')
  }

  const detectedCategories: string[] = []
  const reviewCategories: string[] = []
  let clearlyDetected = false

  for (const [category, score] of Object.entries(scores)) {
    const threshold = thresholds[category]!
    if (score >= threshold) detectedCategories.push(category)
    if (Math.abs(score - threshold) <= reviewMargin) reviewCategories.push(category)
    if (score >= threshold + highConfidenceMargin) clearlyDetected = true
  }

  if (detectedCategories.length >= 2 || clearlyDetected) {
    return {
      riskLevel: 'high',
      riskLevelMessage: 'High model-detected risk',
      detectedCategories,
      reviewCategories,
    }
  }
  if (detectedCategories.length || reviewCategories.length) {
    return {
      riskLevel: 'medium',
      riskLevelMessage: 'Medium model-detected risk',
      detectedCategories,
      reviewCategories,
    }
  }
  return {
    riskLevel: 'low',
    riskLevelMessage: 'Low model-detected risk',
    detectedCategories,
    reviewCategories,
  }
}

File: bert_multilabel/ml_service/risk_level.py
"""Convert multi-label model scores into a simple user-facing risk level.

The model scores are independent, uncalibrated category scores.  This module
therefore compares each score with that category's saved decision threshold;
it never presents a raw score as a legal-risk probability.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any


HIGH_CONFIDENCE_MARGIN = 0.15


def classify_risk_level(
    scores: Mapping[str, float],
    thresholds: Mapping[str, float],
    *,
    review_margin: float = 0.08,
    high_confidence_margin: float = HIGH_CONFIDENCE_MARGIN,
) -> dict[str, Any]:
    """Return a risk-level payload for one clause.

    ``high`` means either two or more categories were detected, or one score
    clearly exceeded its own threshold. ``medium`` means at least one category
    was detected, or a score is close enough to a threshold to review. ``low``
    means no category was detected and none is close to its threshold.
    """
    unknown = set(scores) - set(thresholds)
    if unknown:
        raise ValueError(f"Missing thresholds for categories: {sorted(unknown)}")
    if review_margin < 0 or high_confidence_margin < 0:
        raise ValueError("Risk-level margins must be non-negative")

    detected_categories: list[str] = []
    review_categories: list[str] = []
    clearly_detected_categories: list[str] = []

    for category, score in scores.items():
        threshold = thresholds[category]
        if score >= threshold:
            detected_categories.append(category)
        if abs(score - threshold) <= review_margin:
            review_categories.append(category)
        if score >= threshold + high_confidence_margin:
            clearly_detected_categories.append(category)

    if len(detected_categories) >= 2 or clearly_detected_categories:
        level = "high"
        message = "High model-detected risk"
    elif detected_categories or review_categories:
        level = "medium"
        message = "Medium model-detected risk"
    else:
        level = "low"
        message = "Low model-detected risk"

    return {
        "riskLevel": level,
        "riskLevelMessage": message,
        "detectedCategories": detected_categories,
        "reviewCategories": review_categories,
    }


def classify_score_vector(
    labels: Sequence[str], scores: Sequence[float], thresholds: Mapping[str, float], **kwargs: Any
) -> dict[str, Any]:
    """Convenience wrapper for a model output vector in label order."""
    if len(labels) != len(scores):
        raise ValueError("The label and score vectors must have the same length")
    return classify_risk_level(dict(zip(labels, scores)), thresholds, **kwargs)
